### ISSUE TITLE
Last seen feature not working

### ISSUE THREAD

**Issue by MildPanda** [RipMeApp/ripme#448]
* Ripme version: 1.7.21
* Java version: 
`java version "9.0.4"
Java(TM) SE Runtime Environment (build 9.0.4+11)
Java HotSpot(TM) 64-Bit Server VM (build 9.0.4+11, mixed mode)`
* Operating system: 
Microsoft Windows [Version 10.0.14393]
* Exact URL you were trying to rip when the problem occurred:
N/A
* Please include any additional information about how to reproduce the problem:
Just casually rip anything using the history.end_rip_after_already_seen = # in your properties file
## Expected Behavior
Skip to next download source if # are found to be seen already

## Actual Behavior
Nothing, as if the setting wasn't enabled
Detail the actual (incorrect) behavior here. You can post log snippets or attach log files to your issue report.

It seems like this newer version of ripme doesn't use a properties file inside the folder which ripme is located instead uses a properties inside of the .java file itself which could be replaced with an updated properties file if you just drag the new one into it when using winzip on the ripme file. If not then this may be my issue as I dont see any properties file being created.

**Comment by cyian-1756** [comment]
> It seems like this newer version of ripme doesn't use a properties file inside the folder which ripme is located

The properties file moved several versions ago, on windows it can be found in %APPDATA%\ripme

**Comment by MildPanda** [comment]
Huh, still seems to give issues. I modified the ripme.properties file within AppData\Local (included picture shows current settings) 
![notepad _2018-03-01_20-27-19](https://user-images.githubusercontent.com/5686244/36883414-f9084b90-1d8e-11e8-874c-2e2df96d8af9.png)

**Comment by cyian-1756** [comment]
Did you exit ripme before editting the config file?

What URL where you trying to rip?

**Comment by MildPanda** [comment]
Yes I did exit it beforehand and have restarted the app multiple times while the properties file remains the same of course. In this instance I was ripping from an instagram page plus a reddit user

**Comment by dnecto** [comment]
I can reproduce this for Instagram (at least). The problem is that Instagram will give out different URLs for the same image.

e.g. if you rip an Instagram account once, you get the URL:

https://instagram.fadl1-1.fna.fbcdn.net/5B23ACCD/t51.2885-15/e15/11230326_1593889070893212_545131186_n.jpg

But when you rip it again, the same image will have a different URL:

https://instagram.fadl1-1.fna.fbcdn.net/5B4B39CD/t51.2885-15/e15/11230326_1593889070893212_545131186_n.jpg

Because the history file goes by URL, ripme determines the files are different.

The only real fix for this is to determine uniqueness by filename instead of URL. Of course, this causes other issues relating to potential duplicates.

IMHO, the best bet may be to store an individual history for each album (maybe a `metadata.json` within the album folder?).

This also resolves a separate issue where, if two albums reference the same image, the second one would be considered a duplicate when it really should be downloaded again.

**Comment by cyian-1756** [comment]
@MildPanda Can you post the reddit url you where trying to ripme? Also as @dnecto pointed out it seems that IG changes the url of each image between rips

I should be able to fix this by changing the history bit of the url to Xs (So it ends up looking like https://instagram.fadl1-1.fna.fbcdn.net/XXXXXXXX/t51.2885-15/e15/11230326_1593889070893212_545131186_n.jpg), but that would announce pretty clearly to IG that you were ripping so I want to test and make sure they won't start handing out IP bans 

I'll get on fixing this in the next few days

**Comment by MildPanda** [comment]
I don't know if this was coded in but it should also end rip if its on the last source and it has already seen # in a row as this doesn't seem to be the case currently. 
I am currently just testing the feature out so I am only using 2 seperate reddit urls which are of course **NSFW** and are
https://www.reddit.com/user/gamergal69
https://www.reddit.com/user/chickpeasyx

**Comment by cyian-1756** [comment]
> I don't know if this was coded in but it should also end rip if its on the last source and it has already seen #

You mind explaining that in a bit more detail?

**Comment by MildPanda** [comment]
When you run re-rip checked at the end of queue it should finish ripping once the last source from history has hit your last seen number. So if the ripper is on the most recent rip source (because it rips from old source to new) and there isn't another source to skip to if last seen number is hit then rip = complete and ripme should finish.

**Comment by dnecto** [comment]
> I should be able to fix this by changing the history bit of the url to Xs (So it ends up looking like https://instagram.fadl1-1.fna.fbcdn.net/XXXXXXXX/t51.2885-15/e15/11230326_1593889070893212_545131186_n.jpg), but that would announce pretty clearly to IG that you were ripping so I want to test and make sure they won't start handing out IP bans

Maybe don't request the "Xs" URL; just sanitise it before saving it to the URL history and when checking it on subsequent rips. 

Also, I've just realised it works alright without that part of the URL (e.g. https://instagram.fadl1-1.fna.fbcdn.net/t51.2885-15/e15/11230326_1593889070893212_545131186_n.jpg) so that might be the least conspicuous option

**Comment by cyian-1756** [comment]
@dnecto 

> Maybe don't request the "Xs" URL; just sanitise the real URL before saving it to the URL history and when checking it on subsequent rips.

That is a good idea and is something that should be implemented, but it would require some API changes, tons of testing, changes to the way url history is recorded and would be rather time consuming.  I'm just going to patch this ripper for now and add in a way for rippers to normalize urls before saving them to the history file in the future

@MildPanda 

The changes to the IG ripper be in 1.7.22 which should be out in the next week

**Comment by cyian-1756** [comment]
@MildPanda The instagram issue should be fixed in 1.7.22

**Comment by torbica** [comment]
Maybe I am missing something but I think last seen feature still does not work.
In my debug session your line ` if (alreadyDownloadedUrls >= Utils.getConfigInteger("history.end_rip_after_already_seen", 1000000000) && !isThisATest()) {`
is being hit only once on beginning of rip. It should be after each increase of alreadyDownloadedUrls. I tried to implement this quickly but I do not know what would be best solution for the circuit breaker design pattern.

**Comment by cyian-1756** [comment]
@torbica You mind rewording that? I'm having a bit of trouble following it

**Comment by torbica** [comment]
This feature does not work or I did not actually understand how it should work .
I expected that ripping will stop after x (as defined with history.end_rip_after_already_seen) already downloaded files. This does not happen even with 1.7.22. version. 

I tried to fix the problem but I do not know best way to cancel ripping in process. One of possible issues is that your check `if (alreadyDownloadedUrls >= history.end_rip_after_already_seen` is being called only once. Every time when you increase number of alreadyDownloadedUrls you should do the check and cancel if necessary.

Hope that this explanation is at least little bit better.

**Comment by cyian-1756** [comment]
@torbica Thanks for rewording

> I did not actually understand how it should work

> One of possible issues is that your check if (alreadyDownloadedUrls >= history.end_rip_after_already_seen is being called only once. 

But it's not, as you can see [here](https://github.com/RipMeApp/ripme/blob/a54b71065c6eda9272510b755d8304a4db04c72f/src/main/java/com/rarchives/ripme/ripper/AbstractHTMLRipper.java#L70) it's called in a while loop. That loop fires for every page of an album, meaning that the `if` is checked on every page ripme rips

**Comment by torbica** [comment]
Yes it is in the loop, but [this line](https://github.com/RipMeApp/ripme/blob/a54b71065c6eda9272510b755d8304a4db04c72f/src/main/java/com/rarchives/ripme/ripper/AbstractHTMLRipper.java#L75) will go through the instagram account and collect links instead of going back to your [check](https://github.com/RipMeApp/ripme/blob/a54b71065c6eda9272510b755d8304a4db04c72f/src/main/java/com/rarchives/ripme/ripper/AbstractHTMLRipper.java#L70).  Put breakpoint on your line and you will see it.

**Comment by dnecto** [comment]
To clarify the effect of @torbica's observations:

It seems to be correctly detecting instagram dupes now, based on my testing.

What it's not doing, is stopping after the threshold (`history.end_rip_after_already_seen`) is reached. So if my threshold is set at 5 and the instagram account has 50 images, I'll get 50 lines of "Already downloaded https://instagram.fadl1-1.fna.fbcdn.net/..."

**Comment by cyian-1756** [comment]
@torbica It looks like you're right, last seen will break on any site that collects all likes in one run of `getURLsFromPage`

**Comment by revanwolf** [comment]
> The properties file moved several versions ago, on windows it can be found in %APPDATA%\ripme

You don't remember it normally support both since 1.6.9(portable mode re-included after issue #152 ).
